File: apps/desktop/src-tauri/src/api/fs.rs
use crate::api::ApiClient;
use crate::error::AppResult;
use crate::models::FsTreeEntry;

pub async fn browse_directory(client: &ApiClient, path: Option<&str>) -> AppResult<Vec<FsTreeEntry>> {
    match path {
        Some(p) => client.get_with_query("/fs/browse", &[("path", p)]).await,
        None => client.get("/fs/browse").await,
    }
}

pub async fn pick_folder(client: &ApiClient) -> AppResult<serde_json::Value> {
    client.post("/fs/pick-folder", &serde_json::json!({})).await
}

pub async fn get_directory_tree(
    client: &ApiClient,
    path: Option<&str>,
    depth: Option<u32>,
) -> AppResult<serde_json::Value> {
    let depth_str;
    let mut query: Vec<(&str, &str)> = Vec::new();
    if let Some(p) = path {
        query.push(("path", p));
    }
    if let Some(d) = depth {
        depth_str = d.to_string();
        query.push(("depth", &depth_str));
    }
    if query.is_empty() {
        client.get("/fs/tree").await
    } else {
        client.get_with_query("/fs/tree", &query).await
    }
}

pub async fn read_file(
    client: &ApiClient,
    path: &str,
    start_line: Option<u32>,
    end_line: Option<u32>,
) -> AppResult<serde_json::Value> {
    let start_str;
    let end_str;
    let mut query: Vec<(&str, &str)> = vec![("path", path)];
    if let Some(s) = start_line {
        start_str = s.to_string();
        query.push(("startLine", &start_str));
    }
    if let Some(e) = end_line {
        end_str = e.to_string();
        query.push(("endLine", &end_str));
    }
    client.get_with_query("/fs/file", &query).await
}

pub async fn write_file(client: &ApiClient, path: &str, content: &str) -> AppResult<serde_json::Value> {
    let body = serde_json::json!({ "path": path, "content": content });
    client.post("/fs/file", &body).await
}

pub async fn delete_file(client: &ApiClient, path: &str) -> AppResult<serde_json::Value> {
    client.delete_with_query("/fs/file", &[("path", path)]).await
}

pub async fn create_directory(client: &ApiClient, path: &str) -> AppResult<serde_json::Value> {
    let body = serde_json::json!({ "path": path });
    client.post("/fs/dir", &body).await
}

pub async fn delete_directory(client: &ApiClient, path: &str) -> AppResult<serde_json::Value> {
    client.delete_with_query("/fs/dir", &[("path", path)]).await
}

File: apps/console-macos/Sources/ConsoleCore/Api/Fs.swift
import Foundation

// Mirrors api/fs.rs

public extension ApiClient {
    func browseDirectory(path: String? = nil) async throws -> [FsTreeEntry] {
        var query: [URLQueryItem] = []
        if let path { query.append(.init(name: "path", value: path)) }
        return try await get("/fs/browse", query: query)
    }

    func pickFolder() async throws -> JSONValue {
        try await post("/fs/pick-folder")
    }

    func getDirectoryTree(path: String? = nil, depth: Int? = nil) async throws -> JSONValue {
        var query: [URLQueryItem] = []
        if let path { query.append(.init(name: "path", value: path)) }
        if let depth { query.append(.init(name: "depth", value: String(depth))) }
        return try await get("/fs/tree", query: query)
    }

    func readFile(path: String, startLine: Int? = nil, endLine: Int? = nil) async throws -> JSONValue {
        var query: [URLQueryItem] = [.init(name: "path", value: path)]
        if let s = startLine { query.append(.init(name: "startLine", value: String(s))) }
        if let e = endLine { query.append(.init(name: "endLine", value: String(e))) }
        return try await get("/fs/file", query: query)
    }

    func writeFile(path: String, content: String) async throws -> JSONValue {
        struct Body: Encodable { let path: String; let content: String }
        return try await post("/fs/file", body: Body(path: path, content: content))
    }

    func deleteFile(path: String) async throws -> JSONValue {
        try await delete("/fs/file", query: [.init(name: "path", value: path)])
    }

    func createDirectory(path: String) async throws -> JSONValue {
        struct Body: Encodable { let path: String }
        return try await post("/fs/dir", body: Body(path: path))
    }

    func deleteDirectory(path: String) async throws -> JSONValue {
        try await delete("/fs/dir", query: [.init(name: "path", value: path)])
    }
}
